File: ios/MTGRules/MTGRules/Models/Card.swift
import Foundation

struct Card: Codable, Identifiable, Hashable {
    var id: String { name }
    let name: String
    let manaCost: String?
    let typeLine: String
    let oracleText: String?
    let colors: [String]
    let keywords: [String]
    let legalities: [String: String]
    let imageUri: String?
    let scryfallUri: String?
    let rulings: [String]

    enum CodingKeys: String, CodingKey {
        case name
        case manaCost = "mana_cost"
        case typeLine = "type_line"
        case oracleText = "oracle_text"
        case colors, keywords, legalities
        case imageUri = "image_uri"
        case scryfallUri = "scryfall_uri"
        case rulings
    }
}

struct InteractionQuery: Codable {
    let cardNames: [String]
    let format: String

    enum CodingKeys: String, CodingKey {
        case cardNames = "card_names"
        case format
    }
}

struct InteractionResult: Codable {
    let cards: [Card]
    let rulings: [String]
    let stackNotes: [String]
    let layerNotes: [String]
    let replacementNotes: [String]
    let summary: String

    enum CodingKeys: String, CodingKey {
        case cards, rulings
        case stackNotes = "stack_notes"
        case layerNotes = "layer_notes"
        case replacementNotes = "replacement_notes"
        case summary
    }
}

File: app/models/card.py
from pydantic import BaseModel


class Card(BaseModel):
    name: str
    mana_cost: str | None = None
    type_line: str
    oracle_text: str | None = None
    power: str | None = None
    toughness: str | None = None
    colors: list[str] = []
    keywords: list[str] = []
    legalities: dict[str, str] = {}
    image_uri: str | None = None
    scryfall_uri: str | None = None
    rulings: list[str] = []


class StackItem(BaseModel):
    """An item on the MTG stack (spell or ability)."""
    source_card: str
    description: str
    controller: str = "Player 1"
    is_ability: bool = False
    targets: list[str] = []


class InteractionQuery(BaseModel):
    card_names: list[str]
    format: str = "all"


class InteractionResult(BaseModel):
    cards: list[Card]
    rulings: list[str]
    stack_notes: list[str]
    layer_notes: list[str]
    replacement_notes: list[str]
    summary: str
